### DISCUSSION TITLE
Optional return expression.

### DISCUSSION THREAD

**Discussion by simongoldstone** [dotnet/csharplang#7119] [Language Ideas] [answered]
I often come across code like this:

``` c#

   public void DoSomething(string someInput)
    {
        if (someInput == null)
            return;

        if (someInput == "invalid input")
            return;

        if (AlreadyExists(someInput))
            return;

        // do something only if valid input
    }

```

So lots of defensive code at the start of a method/function.

Perhaps an "optional return" expression would be a useful language feature. It would make code more concise and would allow for certain new strategies:

``` c#

    public void DoSomething(string someInput)
    {
        return? someInput == null;
        return? someInput == "invalid input";
        return? AlreadyExists(someInput);

        // do something only if valid input
    }

```

In these examples, the return would only execute if the expression was true.

Perhaps another extension to this could be:

``` c#

    public string DoSomething(string someInput)
    {
        return? AlreadyExists(someInput) : null;

        return someInput.ToUpper().Trim();
    }

```

In the above example, the `return ? {expression} : {returnVal};` construct would return `returnVal` if `expression` was true.

Could be `returnif` instead of `return?`.

Effectively the return is optional depending on the expression. Perhaps some form of "pattern matching" similar to switch expressions might work.

**Comment by HaloFour** [comment, marked as answer]
See: #2195

**Reply by simongoldstone** [reply]
Hadn't spotted that. Thanks for the reference. Yes that'll do the job well.
